redhat-developer/observability-operator
# feat: allow setting resource requirements for stack components

# Jira
https://issues.redhat.com/browse/MGDSTRM-5559
https://issues.redhat.com/browse/MGDAPI-2713

# Verification
* Checkout this branch
* Install OO on cluster
```
make install
oc new-project observability-operator-system
```
* Run the operator locally
```
make run
```
* Deploy a Observability CR **without** resource requirements
```
cat << EOF | oc create -f -
  apiVersion: observability.redhat.com/v1
  kind: Observability
  metadata:
    name: example
    namespace: observability-operator-system
  spec: 
    configurationSelector:
      matchLabels:
        monitoring-key: middleware
    resyncPeriod: 1h
    selfContained:
      disableDeadmansSnitch: true
      disablePagerDuty: true
      disableRepoSync: true
EOF
```
* Once install completes, verify components have **no** resource values set
```
oc get alertmanager kafka-alertmanager -n observability-operator-system -o json| jq '.spec.resources'
oc get deployment grafana-operator -n observability-operator-system -o json| jq '.spec.template.spec.containers[0].resources'
oc get grafana kafka-grafana -n observability-operator-system -o json| jq '.spec.resources' 
oc get deployment prometheus-operator -n observability-operator-system -o json| jq '.spec.template.spec.containers[0].resources'
oc get prometheus kafka-prometheus -n observability-operator-system -o json| jq '.spec.resources'
```
* Delete CR to uninstall stack
```
oc delete observability example -n observability-operator-system
```
* Deploy a Observability CR **with** resource requirements
```
cat << EOF | oc create -f -
  apiVersion: observability.redhat.com/v1
  kind: Observability
  metadata:
    name: example
    namespace: observability-operator-system
  spec: 
    configurationSelector:
      matchLabels:
        monitoring-key: middleware
    resyncPeriod: 1h
    selfContained:
      alertManagerResourceRequirement:
        limits:
          cpu: 200m
          memory: 400Mi
        requests:
          cpu: 101m
          memory: 201Mi
      disableDeadmansSnitch: true
      disablePagerDuty: true
      disableRepoSync: true
      grafanaOperatorResourceRequirement:
        limits:
          cpu: 202m
          memory: 402Mi
        requests:
          cpu: 103m
          memory: 203Mi
      grafanaResourceRequirement:
        limits:
          cpu: 204m
          memory: 404Mi
        requests:
          cpu: 105m
          memory: 205Mi
      prometheusOperatorResourceRequirement:
        limits:
          cpu: 206m
          memory: 406Mi
        requests:
          cpu: 107m
          memory: 207Mi
      prometheusResourceRequirement:
        limits:
          cpu: 208m
          memory: 408Mi
        requests:
          cpu: 109m
          memory: 209Mi
EOF
```
* Once stack is installed, verify the stack components **does contain** the resource values as specified in the Observability CR
```
oc get alertmanager kafka-alertmanager -n observability-operator-system -o json| jq '.spec.resources'
oc get deployment grafana-operator -n observability-operator-system -o json| jq '.spec.template.spec.containers[0].resources'
oc get grafana kafka-grafana -n observability-operator-system -o json| jq '.spec.resources' 
oc get deployment prometheus-operator -n observability-operator-system -o json| jq '.spec.template.spec.containers[0].resources'
oc get prometheus kafka-prometheus -n observability-operator-system -o json| jq '.spec.resources'
```
